src/wovnio/wovnphp/Headers.php

```diff
@@ -315,4 +337,31 @@ public function getDocumentURI()
 
         return $this->removeLang($url, $this->requestLang());
     }
+
+    public function constructOriginalURL()
+    {
+        $originalURL = $this->protocol . '://' . $this->originalHost . $this->originalPath;
+        if ($this->query) {
+            $originalURL = $originalURL . '?' . $this->query;
+        }
+        return $originalURL;
+    }
+
+    public function computeRedirectUrl()
+    {
+        $cookieLangCode = $this->cookieLang->getCookieLang();
+        $url = $this->urlKeepTrailingSlash;
+        if ($this->store->hasDefaultLangAlias()) {
```

Comment:
This looks exactly the same `buildHrefLang`

```
    private function buildHrefLang($lang_code)
    {
        $url = $this->headers->urlKeepTrailingSlash;

        if ($this->store->hasDefaultLangAlias()) {
            $url = $this->headers->removeLang($url, $this->store->defaultLang());
            $url = Url::addLangCode($url, $this->store, $lang_code, $this->headers);
        } elseif ($lang_code !== $this->store->defaultLang() || $this->store->settings['url_pattern_name'] === 'custom_domain') {
            $url = Url::addLangCode($url, $this->store, $lang_code, $this->headers);
        }
        return htmlentities($url);
    }
```

Then we need to think of a name for this and move this function somewhere.

Reply:
I would leave out the first line and make an `addLangCodeToUrl($url)` function.

Reply:
Actually, it seems like Url.php already has handling for the defaultLang alias. So I wonder if this special case is even needed anymore. Just call `Url.addLangCode`.

Reply:
Also not sure if you need to apply `htmlentities` for your redirect URL. The hreflang does it because it is being written to a HTML attribute.

Reply:
Yep, I'll go for the shortened version with Url::addLangCode.